dora/src/masm/x64.rs

```diff
@@ -418,6 +432,27 @@ impl MacroAssembler {
     }
 
     pub fn int_shl(&mut self, mode: MachineMode, dest: Reg, lhs: Reg, rhs: Reg) {
+        if has_x_ops() {
+            self.int_shlx(mode, dest, lhs, rhs)
+        } else {
+            self.int_shl_c(mode, dest, lhs, rhs)
+        }
+    }
+
+    pub fn int_shlx(&mut self, mode: MachineMode, dest: Reg, lhs: Reg, rhs: Reg) {
+        let x64 = match mode {
+            MachineMode::Int32 => 0,
+            MachineMode::Int64 => 1,
+            _ => unimplemented!(),
+        };
+
+        emit_vex3_rxbm(self, rhs.msb(), 0, lhs.msb(), VEXM_0F38);
+        emit_vex3_wvlp(self, x64, dest.int(), VEXL_Z, VEXP_66);
+        emit_u8(self, 0xF7);
+        emit_modrm(self, 0b11, rhs.and7(), lhs.and7());
+    }
+
+    pub fn int_shl_c(&mut self, mode: MachineMode, dest: Reg, lhs: Reg, rhs: Reg) {
```

Comment:
The idea behind the MacroAssembler is to provide a CPU-independent interface. It might not always be perfect but the functions `int_shl_c`, `int_shrx`, etc. shouldn't be public and/or be moved to x64/asm.rs.